include/zephyr/devicetree.h

```diff
@@ -3017,6 +3017,58 @@
 #define DT_NODE_HAS_STATUS(node_id, status) \
 	DT_NODE_HAS_STATUS_INTERNAL(node_id, status)
 
+/**
+ * @brief Get compatible of a node by index.
+ *
+ * The compatible returned can be used directly in other DT macros that requires
+ * `compat` as argument
+ *
+ * Example devicetree fragment:
+ *
+ * @code{.dts}
+ *     n: node {
+ *             compatible = "vnd,specific-device", "generic-device";
+ *     }
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *     DT_COMPAT_BY_IDX(DT_NODELABEL(n), 0) // vnd_specific_device
+ *     DT_COMPAT_BY_IDX(DT_NODELABEL(n), 1) // generic_device
+ * @endcode
+ *
+ * @param node_id node identifier
+ * @param idx the index to get
+ * @return the compat at idx
+ */
+#define DT_COMPAT_BY_IDX(node_id, idx) DT_STRING_TOKEN_BY_IDX(node_id, compatible, idx)
+
+/**
+ * @brief Get the first compatible of a node
+ *
+ * The compatible returned can be used directly in other DT macros that requires
+ * `compat` as argument
+ *
+ * Example devicetree fragment:
+ *
+ * @code{.dts}
+ *     n: node {
+ *             compatible = "vnd,specific-device", "generic-device";
+ *     }
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *     DT_COMPAT(DT_NODELABEL(n)) // vnd_specific_device
+ * @endcode
+ *
+ * @param node_id node identifier
+ * @return the compat of node_id
+ */
+#define DT_COMPAT(node_id) DT_COMPAT_BY_IDX(node_id, 0)
```

Comment:
Again, I'm not confident that the DT language guarantees the order of compatibles. Can you link to a particular issue or comment on the concrete use-case this is for?

Reply:
> Can you link to a particular issue or comment on the concrete use-case this is for?

There is no existing use case for this yet, I working on #63172 and there might be a need for the software to detect the number of instances (outside of the driver) by [scanning the devicetree](https://github.com/zephyrproject-rtos/zephyr/pull/63172#pullrequestreview-1647110158) without resorting to Kconfig. With this addition it is possible to obtain the number of interrupt controllers by:

```c
DT_NUM_INST_STATUS_OKAY(DT_COMPAT(node_id));
/* or */
DT_NUM_INST_STATUS_OKAY(DT_COMPAT_BY_IDX(node_id, idx));
```

And yes, this code requires the interrupt controller to have only one compatible / the index of the common compatible is known (provided that the indices are guaranteed)